ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodyBackfiller.java

```diff
@@ -0,0 +1,577 @@
+    return getLatestChainHeadSlot()
+        .map(
+            latestSlot ->
+                prepareBatch(
+                        calculateEarliestSlotForBatch(latestSlot, minCustodyPeriodSlot),
+                        latestSlot,
+                        minCustodyPeriodSlot,
+                        true)
+                    .thenCompose(this::executeBatch)
+                    .thenPeek(__ -> isFirstRoundAfterStartup = false))
+        .orElse(SafeFuture.completedFuture(false));
+  }
+
+  private Optional<UInt64> getLatestChainHeadSlot() {
+    return combinedChainDataClient.getRecentChainData().getChainHeads().stream()
+        .map(ProtoNodeData::getSlot)
+        .max(UInt64::compareTo);
+  }
+
+  private SafeFuture<Boolean> prepareAndExecuteBatch(
+      final Optional<UInt64> earliestAvailableCustodySlot, final UInt64 minCustodyPeriodSlot) {
+    if (earliestAvailableCustodySlot.isEmpty()) {
+      // is the db variable is not initialized
+      // we are most likely coming from a checkpoint sync with a fresh DB
+      // let's choose the most recent slot head +1, so that we make sure we cover the head in the
+      // first batch
+      final Optional<UInt64> mostRecentHeadSlotPlus1 =
+          getLatestChainHeadSlot().map(UInt64::increment);
+
+      return eventuallyResetEarliestAvailableCustodySlot(mostRecentHeadSlotPlus1)
+          .thenApply(__ -> false);
+    }
+
+    LOG.debug(
+        "DasCustodyBackfiller: earliestAvailableCustodySlot {}", earliestAvailableCustodySlot);
+
+    if (earliestAvailableCustodySlot.get().isLessThanOrEqualTo(minCustodyPeriodSlot)) {
+
+      return SafeFuture.completedFuture(false);
+    }
+
+    var latestSlotInBatch = earliestAvailableCustodySlot.get().minusMinZero(1);
+    var earliestSlotInBatch =
+        calculateEarliestSlotForBatch(latestSlotInBatch, minCustodyPeriodSlot);
+
+    return prepareBatch(earliestSlotInBatch, latestSlotInBatch, minCustodyPeriodSlot, false)
+        .thenCompose(this::executeBatch);
+  }
+
+  private UInt64 calculateEarliestSlotForBatch(
+      final UInt64 earliestSlot, final UInt64 minCustodyPeriodSlot) {
+    return earliestSlot.minusMinZero(batchSizeInSlots - 1).max(minCustodyPeriodSlot);
+  }
+
+  private SafeFuture<BatchData> prepareBatch(
+      final UInt64 fromSlot,
+      final UInt64 toSlot,
+      final UInt64 minCustodyPeriodSlot,
+      final boolean isHeadsCustodyCheckBatch) {
+    return combinedChainDataClient
+        .getDataColumnIdentifiers(fromSlot, toSlot, UInt64.valueOf(Integer.MAX_VALUE))
+        .thenCombine(
+            retrieveBlocksWithBlobsInRange(fromSlot, toSlot),
+            (dataColumnSlotAndIdentifiers, slotAndBlockRootWithBlobsPresences) ->
+                new BatchData(
+                    fromSlot,
+                    toSlot,
+                    minCustodyPeriodSlot,
+                    custodyGroupCountManager.getCustodyColumnIndices(),
```

Comment:
please add note why we don't need sampling
(because as far as it's finalized data, we consider it proved, so we don't need to verify it, we only need to backfill data to serve network)